### PR TITLE
model of test cases for code snippet question is implemented

### PR DESCRIPTION
**Pull request by Sandipan-Ghosh**
**Fixed Issues**

#21501 #21502 #21503

**Files Changed**

* CodeSnippetQuestionTestCases.cs - To create Database of Test Cases

* TestCaseType.cs - Enum to store type of the Test Case.

* CodeSnippetQuestion.cs - Added a ICollection of type CodeSnippetQuestionTestCases to set one to many relationship.

* question-programming.component.ts - Added a method addTestCases to add test cases of code snippet questions.

* questions-programming.html - implemented the client side. 

* QuestionRepositoryTest.cs - implemented unit testing of test cases

**Checks**

- [x] Naming Conventions
- [x] Server side code comments (proper English, grammar and no spelling mistakes)
- [x] Client side code comments (proper English, grammar and no spelling mistakes)
- [x] Unused variables, methods, blank spaces and name spaces. 
- [x] Optimal Code and code formatting
- [x] Commit History is proper, no merge is done. 
- [x] Commit/pull request messages should be proper to justify your feature
- [ ] All test cases are executed provided by tester.

**Known Issues**

N/A

**Comments**

Addition of Test Cases of code snippet question and unit testing of the same is implemented.

### CODE REVIEW

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [TrappistDbContext.cs#L91](https://github.com/Promact/trappist/blob/4aa701e89290b6b2f6f18461251dba131a52901d/Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs#L91): Use the overloading mechanism instead of the optional parameters. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS2360)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿using Promact.Trappist.DomainModel.Enum;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Promact.Trappist.DomainModel.Models.Question
+{
+    public class CodeSnippetQuestionTestCases
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs]: Please inherit base model

File: Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        public TestCaseType TestCaseType { get; set; }
+
+        [Required]
+        public string TestCaseInput { get; set; }
+        
+        [Required]
+        public string TestCaseOutput { get; set; }
+
+        [Required]
+        public int TestCaseMarks { get; set; }
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs]: marks should be double not int.

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-dashboard/questions-dashboard.html [hunk 3]
```diff
@@ -50,7 +50,7 @@ <h2 class="pull-left">Sections</h2>
                             <div class="qs-container">
                                 <p class="qs-num">Q.{{i+1}}</p>
                                 <div class="qs-text">
-                                    <p class="question" [innerHTML]="question.questionDetail"></p>
+                                    <p [innerHTML]="question.questionDetail"class="question"></p>
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-dashboard/questions-dashboard.html]: there should be one space before class attribute

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-programming/questions-programming.html [hunk 4]
```diff
@@ -73,14 +73,14 @@
                 <div class="form-group">
                     <label class="h4">Test Case(s)</label>
                     <br />
-                    <button type="button" class="btn btn-primary" [hidden]="formControlModel.showTestCase" (click)="formControlModel.showTestCase=true">Add Test Case</button>
-                    <md2-accordion>
+                    <button type="button" class="btn btn-primary" [hidden]="formControlModel.showTestCase" (click)="formControlModel.showTestCase=true;addTestCases()">Add Test Case</button>
+                    <md2-accordion *ngFor="let testCase of testCases;let i = index">
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-programming/questions-programming.html]: significance of i?

Comment (Sandipan-Ghosh) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-programming/questions-programming.html]: here I refers to the index, its significance is to store the number of iterations that it is doing, if its not implemented then while adding new test case i.e. pushing a new object, the data of the previous test case disappears, to fix that issue it is used.

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [TrappistDbContext.cs#L91](https://github.com/Promact/trappist/blob/30580081687ab8485cd30c61974d728cd4f44a65/Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs#L91): Use the overloading mechanism instead of the optional parameters. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS2360)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿using Promact.Trappist.DomainModel.Enum;
+using System.ComponentModel.DataAnnotations;
+
+namespace Promact.Trappist.DomainModel.Models.Question
+{
+    public class CodeSnippetQuestionTestCases:BaseModel
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs]: formatting is not proper.

File: Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        public string TestCaseInput { get; set; }
+        
+        [Required]
+        public string TestCaseOutput { get; set; }
+
+        [Required]
+        public double TestCaseMarks { get; set; }
+
+        public int CodeSnippetQuestionId { get; set; }
+        public virtual CodeSnippetQuestion CodeSnippetQuestion { get; set; }
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Question/CodeSnippetQuestionTestCases.cs]: not properly formatted

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-programming/questions-programming.component.ts [hunk 7]
```diff
@@ -26,7 +28,10 @@ export class QuestionsProgrammingComponent implements OnInit {
     isCategoryReady: boolean;
     isLanguageReady: boolean;
     isFormSubmitted: boolean;
+    isTestCaseAdded: boolean;
     code: any;
+    testCases: CodeSnippetQuestionsTestCases[];
+    TestCaseType: TestCaseType;
```
Comment (Sandipan-Ghosh) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-programming/questions-programming.component.ts]: it is deliberately kept to make it understood that it enables the enum of TestCaseType in the component.

Comment (Sandipan-Ghosh) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-programming/questions-programming.component.ts]: I will put a comment there, will it serve the purpose?

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [TrappistDbContext.cs#L93](https://github.com/Promact/trappist/blob/b87db8194815e83a95cef9f602eae4fb8ed8e2e0/Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs#L93): Use the overloading mechanism instead of the optional parameters. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS2360)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [TrappistDbContext.cs#L93](https://github.com/Promact/trappist/blob/077482f0c14d330772298557a52de44b76ed6e2a/Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs#L93): Use the overloading mechanism instead of the optional parameters. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS2360)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [TrappistDbContext.cs#L93](https://github.com/Promact/trappist/blob/78151d4d3d318af5b65d8d35f1d137d637b38198/Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs#L93): Use the overloading mechanism instead of the optional parameters. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS2360)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [TrappistDbContext.cs#L93](https://github.com/Promact/trappist/blob/a7f01e636182e5e27a0e281c768f43b2aeeb0a6d/Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs#L93): Use the overloading mechanism instead of the optional parameters. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS2360)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**